### PR TITLE
podunavailablebudget proposal

### PR DESCRIPTION
**Pull request by zmberg**
Signed-off-by: liheng.zms <liheng.zms@alibaba-inc.com>

### Ⅰ. Describe what this PR does
- In voluntary disruption scenarios, PodUnavailableBudget can achieve the effect of preventing application disruption or SLA degradation, which greatly improves the high availability of application services.

### CODE REVIEW

File: docs/proposals/20210614-podunavailablebudget.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    - [Program Comparison](#program-comparison)
+  - [User Cases](#user-cases)
+    - [A Sample Case](#a-sample-case)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+
+In kubernetes, powerful workloads (e.g., deployment, statefulset, etc.) are often used to define and manage applications. It not only manages the number of pods, but also provides powerful rolling capabilities such as MaxUnavailable, Partition, etc. However, workload can only help the application reach the desired number of pods as fast as possible, but does not guarantee that the number of available pods will remain at the desired value at any given moment. In the following voluntary disruption scenarios, there are still business disruption and SLA degradation situations:
+1. When the owner deletes the workload by mistake, all the associated pods will be deleted
+2. When upgrading the application version, all pods are restarted at the same time due to the wrong configuration of the rolling upgrade policy, for example MaxUnavailable=0
```
Comment (furykerry) [inline docs/proposals/20210614-podunavailablebudget.md:39]: even if the user correctly set maxUnavailable in the workload, there is still cases where availability is hampered. For example when workload rolling upgrade and voluntary eviction of pods are both running. The key reason we use PUB is maxUnavailable in workload or PDB cannot protect simultaneous modification of pods. So I think we should hight light such user cases, and do not focus on the un-intentional workload deletion or ill-configured workload.

File: docs/proposals/20210614-podunavailablebudget.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+### API Definition
+```
+// PodUnavailableBudgetSpec defines the desired state of PodUnavailableBudget
+type PodUnavailableBudgetSpec struct {
+	// Selector label query over pods managed by the budget
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// TargetReference contains enough information to let you identify an workload for PodUnavailableBudget
+	// Selector and TargetReference are mutually exclusive, TargetReference is priority to take effect
+	TargetReference *TargetReference `json:"targetRef,omitempty"`
```
Comment (furykerry) [inline docs/proposals/20210614-podunavailablebudget.md:55]: please comments about when target reference is required when we already have selector fields

File: docs/proposals/20210614-podunavailablebudget.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+// TargetReference contains enough information to let you identify an workload for PodUnavailableBudget
+type TargetReference struct {
+	// API version of the referent.
+	APIVersion string `json:"apiVersion,omitempty"`
+	// Kind of the referent.
+	Kind string `json:"kind,omitempty"`
+	// Name of the referent.
+	Name string `json:"name,omitempty"`
+}
+```
```
Comment (furykerry) [inline docs/proposals/20210614-podunavailablebudget.md:76]: please explain the pub status fields

File: docs/proposals/20210614-podunavailablebudget.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+### Implement
+This program customizes the PodUnavailableBudget (later referred to as PUB) CRD resource to describe the desired state of the application, and the working mechanism is shown below:
+
+![PodUnavailableBudget](../img/podUnavailableBudget.png)
+
+1. Many voluntary disruption scenarios can be summarized as modifications to Pod resources, which are divided into the following three categories.
+    a) Modification of Pod metadata and spec section
+    b) DeletionPod, i.e. DeletionTimestamp assignment
+    c) Pod eviction
+2. PUB validating is based on the kubernetes admission webhook mechanism, which decides whether to allow this operation to be executed based on the UnavailableAllowed in the PUB Status:
```
Comment (furykerry) [inline docs/proposals/20210614-podunavailablebudget.md:87]: can we elaborate the data manipulation of pub status here ?

**Review by furykerry — CHANGES_REQUESTED** [review]

File: docs/proposals/20210614-podunavailablebudget.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+PDB limits the number of Pods of a replicated application that are down simultaneously from voluntary disruptions. However, it can only constrain the voluntary disruption triggered by the [Eviction API](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api).
+For example, when you run kubectl drain, the tool tries to evict all of the Pods on the Node you're taking out of service.
+
+In the following voluntary disruption scenarios, there are still business disruption or SLA degradation situations:
+1. The application owner update deployment's pod template for general upgrading, while cluster administrator drain nodes to scale the cluster down(learn about [Cluster Autoscaling](https://github.com/kubernetes/autoscaler/#readme)).
+2. The middleware team is using [SidecarSet](https://openkruise.io/en-us/docs/sidecarset.html) to rolling upgrade the sidecar containers of the cluster, e.g. ServiceMesh envoy, while HPA triggers the scale-down of business applications.
```
Comment (furykerry) [inline docs/proposals/20210614-podunavailablebudget.md:46]: plz add the case for pod eviction initiated by admin, pod eviction is a more general case which occurs even if no kruise workload is used.

File: docs/proposals/20210614-podunavailablebudget.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+	// status information is valid only if observedGeneration equals to PUB's object generation.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration"`
+
+	// DisruptedPods contains information about pods whose eviction or deletion was
+	// processed by the API handler but has not yet been observed by the PodUnavailableBudget.
+	// +optional
+	DisruptedPods map[string]metav1.Time `json:"disruptedPods,omitempty"`
+
+	// UnavailablePods contains information about pods whose specification changed(inplace-update pod),
```
Comment (furykerry) [inline docs/proposals/20210614-podunavailablebudget.md:96]: specification changes does not necessarily means inplace-update, pod container restart is also possible. plz changes to 
(e.g. inplace-update)

File: docs/proposals/20210614-podunavailablebudget.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+// PodUnavailableBudgetStatus defines the observed state of PodUnavailableBudget
+type PodUnavailableBudgetStatus struct {
+	// Most recent generation observed when updating this PUB status. UnavailableAllowed and other
+	// status information is valid only if observedGeneration equals to PUB's object generation.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration"`
+
+	// DisruptedPods contains information about pods whose eviction or deletion was
+	// processed by the API handler but has not yet been observed by the PodUnavailableBudget.
```
Comment (furykerry) [inline docs/proposals/20210614-podunavailablebudget.md:92]: > has not yet been observed by the PodUnavailableBudget
has not yet been observed by the PodUnavailableBudget controller

**Review by furykerry — CHANGES_REQUESTED** [review]

**Review by furykerry — COMMENTED** [review]
LGTM

**Comment by Fei-Guo** [comment]
LGTM

**Review by furykerry — APPROVED** [review]

**Pull request merged**